compiler/rustc_mir/src/const_eval/eval_queries.rs

```diff
@@ -162,14 +162,15 @@ pub(super) fn op_to_const<'tcx>(
         Err(imm) => match *imm {
             Immediate::Scalar(x) => match x {
                 ScalarMaybeUninit::Scalar(s) => ConstValue::Scalar(s),
-                ScalarMaybeUninit::Uninit => to_const_value(&op.assert_mem_place(ecx)),
+                ScalarMaybeUninit::Uninit => to_const_value(&op.assert_mem_place()),
             },
             Immediate::ScalarPair(a, b) => {
-                let (data, start) = match a.check_init().unwrap() {
-                    Scalar::Ptr(ptr) => {
-                        (ecx.tcx.global_alloc(ptr.alloc_id).unwrap_memory(), ptr.offset.bytes())
+                // We know `offset` is relative to the allocation, so we can use `into_parts`.
+                let (data, start) = match ecx.scalar_to_ptr(a.check_init().unwrap()).into_parts() {
```

Comment:
could into_parts assert this flag internally? Or does miri also have a need for it?

Reply:
Yeah Miri needs to access the `offset` to compare between `Pointer<AllocId>` (address is relative) and `Pointer<Tag>` (address is absolute).

Reply:
FWIW, I actually did think of a way to make this more type-safe: the `Provenance` trait could have an associated type `Offset`, and a function `wrap_offset(Size) -> Offset`. Then `into_parts` would return `Offset`, so code generic in `Tag` would actually not know the type `Offset` and this could not use it wrong.

Then we might go even further and make the `Pointer::offset` field be of type `Offset`... though at that point we cannot define generic pointer arithmetic methods any more so we'd need some more methods in the `Provenance` trait.

But I am not sure if that's worth it. I had one bug caused by mis-interpreting a relative offset as absolute, and this would not have caught it (since the bug was about pointers stored inside an `Allocation`, where the offset is part of the byte array anyway).